mods/page_revisions.js

```diff
@@ -293,56 +425,72 @@ PRS.prototype.getTitleRevision = function(restbase, req) {
+                        title: title,
+                        rev: parseInt(rp.revision)
+                    },
+                    limit: 1
+                }
+            })
+            .catch(function(e) {
+                if (e.status !== 404) {
+                    throw e;
+                }
+                return self.fetchAndStoreMWRevision(restbase, req);
+            }),
+            pageData: pageDataRequest
         });
     } else if (!rp.revision) {
         if (req.headers && /no-cache/.test(req.headers['cache-control'])) {
-            revisionRequest = self.fetchAndStoreMWRevision(restbase, req)
+            revisionRequest = P.props({
+                revisionInfo: self.fetchAndStoreMWRevision(restbase, req),
+                pageData: pageDataRequest
+            })
+            .then(function(res) {
+                return self._checkPageUndeleted(restbase, req, res);
+            })
             .catch(function(e) {
                 if (e.status !== 404) {
                     throw e;
                 }
-                return getLatestTitleRevision()
-                // In case 404 is returned by MW api, the page is deleted
-                .then(function(result) {
-                    result = result.body.items[0];
-                    result.tid = uuid.now().toString();
-                    result.restrictions = result.restrictions || [];
-                    result.restrictions.push('page_deleted');
-                    return restbase.put({
-                        uri: self.tableURI(rp.domain),
-                        body: {
-                            table: self.tableName,
-                            attributes: result
-                        }
-                    })
-                    .then(function() {
-                        throw e;
-                    });
+                // Page data request never returns 404, so it's from MW API => page was deleted
+                return self._signalPageDeleted(restbase, req)
+                .then(function() {
+                    throw e;
                 });
             });
         } else {
-            revisionRequest = getLatestTitleRevision()
-            .catch(function(e) {
-                if (e.status !== 404) {
-                    throw e;
+            revisionRequest = P.props({
+                revisionInfo: getLatestTitleRevision()
+                .catch(function(e) {
+                    if (e.status !== 404) {
+                        throw e;
+                    }
+                    return self.fetchAndStoreMWRevision(restbase, req);
+                }),
+                pageData: pageDataRequest
+            })
+            .then(function(res) {
+                if (res.pageData
+                        && res.pageData.body.items
+                        && res.pageData.body.items.length) {
+                    var latestEvent = res.pageData.body.items[0];
+                    if (latestEvent.event_type === 'rename_to') {
```

Comment:
@d00rman Regarding redirect vs 404, there's been a discussion [here](https://phabricator.wikimedia.org/T87393). The main concern was about clients which follow redirects automatically, and wouldn't notice it. That discussion is still unfinished, but I agree that a redirect would be nicer here.

Regarding pointing to the latest title, it's a hard task (both programmatically and computationally), but I think that indeed it would be more correct. We are asking about the latest data, so we need to provide the latest name..